core/src/main/java/novoda/lib/sqliteprovider/util/DatabaseStructure.java

```diff
@@ -0,0 +1,226 @@
+        return columnName.endsWith("_id");
+    }
+
+    public Map<String, SQLiteType> columns(String table) {
+        final Cursor columnsCursor = queryTableColumnsFor(table);
+        Map<String, SQLiteType> columns = parseColumns(columnsCursor);
+        columnsCursor.close();
+        return Collections.unmodifiableMap(columns);
+    }
+
+    private Map<String, SQLiteType> parseColumns(Cursor columnsCursor) {
+        Map<String, SQLiteType> columns = new HashMap<>(columnsCursor.getCount());
+        while (columnsCursor.moveToNext()) {
+            String name = columnsCursor.getString(columnsCursor.getColumnIndexOrThrow(COLUMN_NAME));
+            String type = columnsCursor.getString(columnsCursor.getColumnIndexOrThrow(COLUMN_TYPE));
+            columns.put(name, SQLiteType.fromName(type));
+        }
+        return columns;
+    }
+
+    private Cursor queryTableColumnsFor(String table) {
+        return database.rawQuery(String.format(PRAGMA_TABLE_INFO, table), null);
+    }
+
+    public List<Constraint> uniqueConstraints(String table) {
+        uniqueConstraints.put(table, new ArrayList<Constraint>());
+        lookForImplicitUniqueIndex(table);
+        final Cursor indexCursor = queryIndexListForTable(table);
+        parseUniqueConstraints(indexCursor, table);
+        indexCursor.close();
+        return Collections.unmodifiableList(uniqueConstraints.get(table));
+    }
+
+    private void lookForImplicitUniqueIndex(String table) {
+        // This is an implicit unique index, and won't show up querying the other indexes
+        Constraint integerPrimaryKeyConstraint = findIntegerPrimaryKeyConstraint(table);
+        if (integerPrimaryKeyConstraint != null) {
+            uniqueConstraints.get(table).add(integerPrimaryKeyConstraint);
+        }
+    }
+
+    private void parseUniqueConstraints(Cursor indexCursor, String forTable) {
+        while (indexCursor.moveToNext()) {
+            addUniqueIndexConstraint(indexCursor, forTable);
+        }
+    }
+
+    private void addUniqueIndexConstraint(Cursor indexCursor, String forTable) {
+        if (!isIndexUnique(indexCursor)) {
+            return;
+        }
+        String indexName = indexCursor.getString(1);
+        final Cursor indexInfoCursor = queryIndexInfo(indexName);
+        List<String> columns = parseIndexColumns(indexInfoCursor);
+        indexInfoCursor.close();
+        uniqueConstraints.get(forTable).add(new Constraint(columns));
+    }
+
+    private boolean isIndexUnique(Cursor indexCursor) {
+        return indexCursor.getInt(2) == 1;
+    }
+
+    private List<String> parseIndexColumns(Cursor indexInfoCursor) {
+        List<String> columns = new ArrayList<>(indexInfoCursor.getCount());
+        while (indexInfoCursor.moveToNext()) {
+            String columnName = indexInfoCursor.getString(2);
+            columns.add(columnName);
+        }
+        return columns;
+    }
+
+    private Constraint findIntegerPrimaryKeyConstraint(String table) {
```

Comment:
This should be right after `lookForImplicitUniqueIndex()`, along with its dependencies